lib/Scheduler/graph_trans.cpp

```diff
@@ -2,11 +2,70 @@
 #include "opt-sched/Scheduler/bit_vector.h"
 #include "opt-sched/Scheduler/logger.h"
 #include "opt-sched/Scheduler/register.h"
+#include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
 #include <list>
+#include <vector>
 
 using namespace llvm::opt_sched;
 
+// Gets the registers which may be lengthened by scheduling nodeB after nodeA,
+// as compared to if nodeB was scheduled before nodeA
+static llvm::SmallVector<const Register *, 10> possiblyLengthenedIfAfterOther(
+    const SchedInstruction *nodeB, llvm::ArrayRef<const Register *> bUses,
+    const SchedInstruction *nodeA, llvm::ArrayRef<const Register *> aUses) {
+  llvm::SmallVector<const Register *, 10> result;
+
+  llvm::copy_if(bUses, std::back_inserter(result), [&](const Register *useB) {
+    // Is this register also used by A?
+    // If so, reordering A and B would have no effect on this register's
+    // live range.
+    const bool usedByA =
+        llvm::any_of(aUses, [&](const Register *useA) { return useA == useB; });
+    // If this register isn't used by A, is it at least used
+    // by some successor? If so, reordering A and B would have no effect on
+    // this register's live range, as it must live until C.
+    const auto usedByC = [&] {
+      return llvm::any_of(
+          useB->GetUseList(), [&](const SchedInstruction *user) {
+            // Given: [... B ... A ...]
+            // We need to prove that the register `useB` won't be used by an
+            // instruction before A but after B. In the hypothetical schedule we
+            // are considering, A currently appears after B. Thus, it is
+            // sufficient to show that this register has a user C that is a
+            // successor of A.
+            //
+            // This is more relaxed than showing that C is a successor of B, as
+            // RcrsvScsr(B) is a subset of RcrsvScsr(A).
+            return user != nodeB &&
+                   nodeA->IsRcrsvScsr(const_cast<SchedInstruction *>(user));
+          });
+    };
+
+    return !usedByA && !usedByC();
+  });
+
+  return result;
+}
+
+// Gets the Uses for the given SchedInstruction.
+static llvm::ArrayRef<const Register *> getUses(SchedInstruction *node) {
+  Register **uses;
+  const int useCount = node->GetUses(uses);
+  assert(useCount >= 0);
+  return {uses, static_cast<size_t>(useCount)};
```

Comment:
Just wondering, does makeArrayRef work here?

Reply:
Indeed it would. It calls the same constructor that I'm calling here.